ceremonies/2023/run.sh

```diff
@@ -0,0 +1,110 @@
+#!/bin/bash -e
+
+set -o pipefail
+
+function usage() {
+    echo -e "Usage:
+
+    ./$(basename "${0}")
+    "
+}
+
+if [ "${1}" == "-h" ]; then
+    usage
+    exit 0
+fi
+
+function _echo() {
+    echo "$(date +'%Y/%m/%d %H:%M:%S') ${@}"
+}
+
+function setup_ceremony_tool() {
+    if [ -n "${CEREMONY_BIN_2023}" ] && [ -x "${CEREMONY_BIN_2023}" ]; then
+        return 0
+    fi
+
+    TMPDIR="/tmp/ceremony-tools"
+    export CEREMONY_BIN_2023="${TMPDIR}/bin/2023/ceremony"
+    mkdir -p "${TMPDIR}/bin/2023/"
+    if [ ! -d "${TMPDIR}/boulder" ]; then
+        git clone https://github.com/letsencrypt/boulder/ "${TMPDIR}/boulder"
+    fi
+
+    if [ ! -x "${TMPDIR}/bin/2023/ceremony" ]; then
+        # Build ceremony on the commit prior to removing configuration of Policy OIDs.
+        # This will allow all ceremonies prior to 2023 to complete successfully without
+        # requiring backporting changes to those ceremonies and losing the historical
+        # representation of the ceremony.
+        cd "${TMPDIR}/boulder"
+        git checkout 72e01b337abccc9b849a3063666943489bcc573d
+        make
+        cd -
+        cp "${TMPDIR}/boulder/bin/ceremony" "${TMPDIR}/bin/2023/"
+    fi
+}
+
+setup_ceremony_tool
+
+CEREMONY_BIN="${CEREMONY_BIN_2023}"
+CEREMONY_YEAR="$(basename "$(dirname "$(readlink -f "${0}")")")"
+_echo "Running ${CEREMONY_YEAR} ceremony with tooling at ${CEREMONY_BIN}"
+
+CEREMONY_DIR="$(dirname ${BASH_SOURCE[0]})"
+cd "${CEREMONY_DIR}"
+
+"${CEREMONY_BIN}" --config "./e5-key.yaml"
+"${CEREMONY_BIN}" --config "./e6-key.yaml"
+"${CEREMONY_BIN}" --config "./e7-key.yaml"
+"${CEREMONY_BIN}" --config "./e8-key.yaml"
+"${CEREMONY_BIN}" --config "./e9-key.yaml"
+"${CEREMONY_BIN}" --config "./r10-key.yaml"
+"${CEREMONY_BIN}" --config "./r11-key.yaml"
+"${CEREMONY_BIN}" --config "./r12-key.yaml"
+"${CEREMONY_BIN}" --config "./r13-key.yaml"
+"${CEREMONY_BIN}" --config "./r14-key.yaml"
+
+"${CEREMONY_BIN}" --config "./e5-cert.yaml"
+"${CEREMONY_BIN}" --config "./e6-cert.yaml"
+"${CEREMONY_BIN}" --config "./e7-cert.yaml"
+"${CEREMONY_BIN}" --config "./e8-cert.yaml"
+"${CEREMONY_BIN}" --config "./e9-cert.yaml"
+"${CEREMONY_BIN}" --config "./r10-cert.yaml"
+"${CEREMONY_BIN}" --config "./r11-cert.yaml"
+"${CEREMONY_BIN}" --config "./r12-cert.yaml"
+"${CEREMONY_BIN}" --config "./r13-cert.yaml"
+"${CEREMONY_BIN}" --config "./r14-cert.yaml"
+
+"${CEREMONY_BIN}" --config "./e5-cross-cert.yaml"
+"${CEREMONY_BIN}" --config "./e6-cross-cert.yaml"
+"${CEREMONY_BIN}" --config "./e7-cross-cert.yaml"
+"${CEREMONY_BIN}" --config "./e8-cross-cert.yaml"
+"${CEREMONY_BIN}" --config "./e9-cross-cert.yaml"
+
+
+## 1704067201 is Dec 31, 2024; this ensures the check will still pass even after root-x1 and root-x2 expire.
```

Comment:
root-x1 and root-x2 don't expire before Dec 31 2024. I think this means it's ensuring that it still works after the cross-signs from DST Root CA X3 expire?

Reply:
I've removed the second half of the comment. That verbiage was leftover from the original run script. What I've since found out from [the openssl docs](https://www.openssl.org/docs/man3.0/man1/openssl-verification-options.html) is
> For each element in the chain, including the root CA certificate, the validity period as specified by the notBefore and notAfter fields is checked against the current system time. The -attime flag may be used to use a reference time other than "now." The certificate signature is checked as well (except for the signature of the typically self-signed root CA certificate, which is verified only if the -check_ss_sig option is given).

What the comment actually meant was that verification of the chain would take place at the provided time, rather than in the future when some part of the chain may have expired.